### ISSUE TITLE
POSIX::strftime hangs with pattern %z

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10185]
Migrated from [rt.perl.org#72928](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72928) (status was 'resolved')

Searchable as RT72928$

**Comment by p5pRT** [comment]
### From pavel.krejci@siemens-enterprise.com

#### Created by pavel.krejci@siemens-enterprise.com



The POSIX&#8203;::strftime hangs when the pattern %z is specified&#8203;:

POSIX&#8203;::tzset\(\);
my \($sec\, $min\, $hour\, $day\, $mon\, $year\, $wday\,$yday\,$isdst\) = localtime\( time\(\) \) ;
my $tz\_off = POSIX&#8203;::strftime\( "%z"\, $sec\,$min\,$hour\,$day\,$mon\,$year\);

The strace shows that perl makes infinite loop&#8203;:
stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
\.\.\.

My Linux is SLES10 SP3&#8203;:
linux&#8203;:\~ \# uname \-a
Linux linux 2\.6\.16\.60\-0\.54\.5\-default \#1 Fri Sep 4 01&#8203;:28&#8203;:03 UTC 2009 i686 i686 i386 GNU/Linux

linux&#8203;:\~ \# rpm \-qi glibc
Name        : glibc                        Relocations&#8203;: \(not relocatable\)
Version     : 2\.4                               Vendor&#8203;: SUSE LINUX Products GmbH\, Nuernberg\, Germany
Release     : 31\.74\.1                       Build Date&#8203;: Fri Sep  4 20&#8203;:25&#8203;:36 2009

Thanks
Pavel Krejci

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 - Fri Jul 11 14:53:55 UTC 2008
It is being executed now by  Perl v5.8.8 - Fri Jul 11 14:50:58 UTC 2008.

Site configuration information for perl v5.8.8:

Configured by abuild at Fri Jul 11 14:50:58 UTC 2008.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.16, archname=i586-linux-thread-multi
    uname='linux cara 2.6.16 #1 smp thu may 17 14:00:09 utc 2007 i686 i686 i386 gnulinux '
    config_args='-ds -e -Dprefix=/usr -Dvendorprefix=/usr -Dinstallusrbinperl -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Duseshrplib=true -Doptimize=-O2 -march=i586 -mtune=i686 -fmessage-length=0 -Wall -D_FORTIFY_SOURCE=2 -g -Wall -pipe'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -pipe -Wdeclaration-after-statement -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=i586 -mtune=i686 -fmessage-length=0 -Wall -D_FORTIFY_SOURCE=2 -g -Wall -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -pipe -Wdeclaration-after-statement'
    ccversion='', gccversion='4.1.2 20070115 (prerelease) (SUSE Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =''
    libpth=/lib /usr/lib /usr/local/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.4.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.8.8/i586-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared'

Locally applied patches:



@INC for perl v5.8.8:
    /usr/lib/perl5/5.8.8/i586-linux-thread-multi
    /usr/lib/perl5/5.8.8
    /usr/lib/perl5/site_perl/5.8.8/i586-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.8/i586-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.8:
    HOME=/root
    LANG=POSIX
    LANGUAGE (unset)
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/local/sbin:/opt/gnome/sbin:/root/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/games:/opt/bin:/opt/gnome/bin:/usr/lib/mit/bin:/usr/lib/mit/sbin:/opt/bin:/opt/bin:/opt/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Feb 18\, 2010 at 08&#8203;:42&#8203;:24AM \-0800\, Krejci\, Pavel wrote&#8203;:
> The POSIX&#8203;::strftime hangs when the pattern %z is specified&#8203;:
> 
> POSIX&#8203;::tzset\(\);
> my \($sec\, $min\, $hour\, $day\, $mon\, $year\, $wday\,$yday\,$isdst\) = localtime\( time\(\) \) ;
> my $tz\_off = POSIX&#8203;::strftime\( "%z"\, $sec\,$min\,$hour\,$day\,$mon\,$year\);
> 
> The strace shows that perl makes infinite loop&#8203;:
> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0

This appears to have been fixed in 5\.8\.9 and 5\.10\.0

\-\- 
It's not that I'm afraid to die\, I just don't want to be there when it
happens\.
&nbsp;    \-\- Woody Allen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Citeren Dave Mitchell \<davem@&#8203;iabyn\.com>&#8203;:

> On Thu\, Feb 18\, 2010 at 08&#8203;:42&#8203;:24AM \-0800\, Krejci\, Pavel wrote&#8203;:
>> The POSIX&#8203;::strftime hangs when the pattern %z is specified&#8203;:
>>
>> POSIX&#8203;::tzset\(\);
>> my \($sec\, $min\, $hour\, $day\, $mon\, $year\, $wday\,$yday\,$isdst\) =   
>> localtime\( time\(\) \) ;
>> my $tz\_off = POSIX&#8203;::strftime\( "%z"\, $sec\,$min\,$hour\,$day\,$mon\,$year\);
>>
>> The strace shows that perl makes infinite loop&#8203;:
>> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
>> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
>> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
>> stat64\("/etc/localtime"\, \{st\_mode=S\_IFREG|0644\, st\_size=2246\, \.\.\.\}\) = 0
>
> This appears to have been fixed in 5\.8\.9 and 5\.10\.0

See also&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=52126
\#52126&#8203;: POSIX&#8203;::strftime hangs on %Z under cygwin

which was fixed with commit \#29350&#8203;:
http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/7743c30757199ca2f38bf73a40c6d3a6aef6b5ea

author    Steve Hay \<SteveHay@&#8203;planit\.com>
&nbsp;    Wed\, 22 Nov 2006 15&#8203;:11&#8203;:41 \+0000 \(15&#8203;:11 \+0000\)
committer    Steve Hay \<SteveHay@&#8203;planit\.com>
&nbsp;    Wed\, 22 Nov 2006 15&#8203;:11&#8203;:41 \+0000 \(15&#8203;:11 \+0000\)
commit    7743c30757199ca2f38bf73a40c6d3a6aef6b5ea
tree    69776f05ffc80209070837f966722e8f3fd5da7e    tree | snapshot
parent    a870056241d074c8220de8211d965e13ac2d3c7d    commit | diff

Fix infinite loop in Perl\_my\_strftime\(\) for failing strftime\(\)

p4raw\-id&#8203;: //depot/perl@&#8203;29350


Best regards\,

Bram

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
